photon-lib/src/main/java/org/photonvision/RobotPoseEstimator.java

```diff
@@ -41,300 +41,461 @@
+    }
+
+    /**
+     * Poll data from the configured cameras and update the estimated position of the robot. Returns
+     * empty if there are no cameras set or no targets were found from the cameras.
+     *
+     * @return a Pair of the estimated Position of the robot in the form of a Pose3d object and the
+     *     timestamp of that position from the camera in seconds that is comparable to the FPGA
+     *     hardware clock (in seconds).
      */
     public Optional<Pair<Pose3d, Double>> update() {
         if (cameras.isEmpty()) {
             DriverStation.reportError("[RobotPoseEstimator] Missing any camera!", false);
             return Optional.empty();
         }
 
-        Pair<Pose3d, Double> pair = getResultFromActiveStrategy();
+        ArrayList<Pair<PhotonPipelineResult, Transform3d>> cameraResults =
+                new ArrayList<>(cameras.size());
+        boolean hasTargets = false;
 
-        if (pair != null) {
-            lastPose = pair.getFirst();
+        for (Pair<PhotonCamera, Transform3d> p : cameras) {
+            PhotonPipelineResult pipelineResults = p.getFirst().getLatestResult();
+
+            if (pipelineResults.hasTargets()) hasTargets = true;
+
+            cameraResults.add(Pair.of(pipelineResults, p.getSecond()));
         }
 
-        return Optional.ofNullable(pair);
-    }
+        if (!hasTargets) {
+            return Optional.empty();
+        }
 
-    private Pair<Pose3d, Double> getResultFromActiveStrategy() {
+        Optional<Pair<Pose3d, Double>> estimatedPose;
         switch (strategy) {
             case LOWEST_AMBIGUITY:
-                return lowestAmbiguityStrategy();
+                estimatedPose = lowestAmbiguityStrategy(cameraResults);
+                break;
             case CLOSEST_TO_CAMERA_HEIGHT:
-                return closestToCameraHeightStrategy();
-            case CLOSEST_TO_REFERENCE_POSE:
-                return closestToReferencePoseStrategy();
+                estimatedPose = closestToCameraHeightStrategy(cameraResults);
+                break;
             case CLOSEST_TO_LAST_POSE:
-                return closestToLastPoseStrategy();
+                setReferencePose(lastPose);
+            case CLOSEST_TO_REFERENCE_POSE:
+                estimatedPose = closestToReferencePoseStrategy(cameraResults, referencePose);
+                break;
             case AVERAGE_BEST_TARGETS:
-                return averageBestTargetsStrategy();
+                estimatedPose = averageBestTargetsStrategy(cameraResults);
+                break;
             default:
-                DriverStation.reportError("[RobotPoseEstimator] Invalid pose strategy!", false);
-                return null;
+                DriverStation.reportError(
+                        "[RobotPoseEstimator] Unknown Position Estimation Strategy!", false);
+                return Optional.empty();
+        }
+
+        if (estimatedPose.isEmpty()) {
+            lastPose = null;
```

Comment:
`lastPose` never gets set if it is present. Also, should `getLastPose()` return an optional, or should it always return the last pose generated? That way, if `update()` returns an empty optional, users can know where the estimator previously thought the robot was. I'm sort of leaning towards returning an optional, but the latter might be handy for some teams.

Reply:
I would like to remove a lot of stuff from this class, settings field after the constructor can be removed, and `SetCamera` should be removed from the c++ side, it only exists to fulfill tests. The field will likely not have dynamically positioned tags and people are more likely to just add all cameras in the constructor. The way it currently supports bad programming practices.

Reply:
I had to rush most of the class, so many things need to be improved & changed around. The original PR wasn't really ready for merge, but we wanted to get it out before the season to find bugs & improve it. There is a lot of refactoring that I wasn't able to get to (and that you've already done; thanks!); a lot of the logic is copied among the strategies. I'm a Java programmer, so the C++ code should be quite stinky, so feel free to make large changes to that haha.
Optionals would work fine, but would require more work on the implementation side for teams.
Regarding removing the setters, I am of the opinion that if a case exists for a feature exists, we should keep it. Some bad practices can occur from the end-user, but it is better than an end-user having to roll their own build IMO.

Reply:
I also suck at c++ lol but will give it a go. If the setters only exist for future use, they should be added then. RN, it will appear confusing to the end user. Also, I feel like optional will be better for people who are solely really on IDE. This way is better than people throwing null pointers and not understanding why,